File: zilli/models/llamacpp.py
from __future__ import annotations

import json
import logging
import time
from typing import AsyncIterator

from zilli.models.base import GenerationResult, ModelBackend

logger = logging.getLogger("zilli.models.llamacpp")

try:
    import httpx
    HAS_HTTPX = True
except ImportError:
    HAS_HTTPX = False
    httpx = None


class LlamaCppBackend(ModelBackend):
    def __init__(self, name: str, model_id: str, base_url: str = "http://127.0.0.1:8080"):
        super().__init__(name, model_id, base_url)
        self._completion_url = f"{self.base_url}/completion"
        self._health_url = f"{self.base_url}/health"
        self._client: httpx.AsyncClient | None = None

    def _get_client(self) -> httpx.AsyncClient | None:
        if self._client is None and HAS_HTTPX:
            self._client = httpx.AsyncClient(timeout=httpx.Timeout(300.0))
        return self._client

    async def generate(
        self,
        prompt: str,
        max_tokens: int = 2048,
        temperature: float = 0.1,
    ) -> GenerationResult:
        if not HAS_HTTPX:
            return GenerationResult(
                text="", model_name=self.model_id,
                error="httpx not installed",
            )

        client = self._get_client()
        if client is None:
            return GenerationResult(
                text="", model_name=self.model_id,
                error="Failed to create HTTP client",
            )

        payload = {
            "prompt": prompt,
            "n_predict": max_tokens,
            "temperature": temperature,
            "cache_prompt": True,
        }

        start = time.monotonic()
        try:
            response = await client.post(self._completion_url, json=payload)
            duration_ms = (time.monotonic() - start) * 1000

            if response.status_code != 200:
                return GenerationResult(
                    text="", model_name=self.model_id,
                    error=f"HTTP {response.status_code}: {response.text[:200]}",
                    duration_ms=duration_ms,
                )

            data = response.json()
            return GenerationResult(
                text=data.get("content", ""),
                model_name=self.model_id,
                tokens_in=data.get("tokens_cached", 0) or data.get("tokens_evaluated", 0),
                tokens_out=data.get("tokens_predicted", 0),
                duration_ms=duration_ms,
            )
        except Exception as e:
            duration_ms = (time.monotonic() - start) * 1000
            return GenerationResult(
                text="", model_name=self.model_id,
                error=str(e), duration_ms=duration_ms,
            )

    async def generate_stream(
        self,
        prompt: str,
        max_tokens: int = 2048,
        temperature: float = 0.1,
    ) -> AsyncIterator[str]:
        if not HAS_HTTPX:
            yield ""
            return

        client = self._get_client()
        if client is None:
            yield ""
            return

        payload = {
            "prompt": prompt,
            "n_predict": max_tokens,
            "temperature": temperature,
            "cache_prompt": True,
            "stream": True,
        }

        try:
            async with client.stream("POST", self._completion_url, json=payload) as response:
                if response.status_code != 200:
                    logger.error("llama.cpp stream HTTP %d", response.status_code)
                    yield ""
                    return
                async for line in response.aiter_lines():
                    if not line.strip():
                        continue
                    try:
                        data = json.loads(line)
                        chunk = data.get("content", "")
                        if chunk:
                            yield chunk
                        if data.get("stop", False):
                            break
                    except json.JSONDecodeError:
                        continue
        except Exception as e:
            logger.error("llama.cpp stream error: %s", e)
            yield ""

    async def health_check(self) -> bool:
        if not HAS_HTTPX:
            return False
        try:
            client = self._get_client()
            if client is None:
                return False
            resp = await client.get(self._health_url, timeout=httpx.Timeout(5.0))
            return resp.status_code == 200
        except Exception as e:
            logger.warning("llama.cpp health check failed: %s", e)
            return False

File: zilli-rs/src/models/backends/llamacpp.rs
use async_trait::async_trait;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::{Duration, Instant};
use tracing::instrument;

use super::{GenerationResult, ModelBackend, ModelError};

#[derive(Debug)]
pub struct LlamaCppBackend {
    name: String,
    model_id: String,
    base_url: String,
    client: Client,
}

#[derive(Serialize)]
struct CompletionRequest {
    prompt: String,
    n_predict: Option<i32>,
    temperature: Option<f64>,
    stream: bool,
}

#[derive(Deserialize)]
struct CompletionResponse {
    content: String,
    timings: Option<Timings>,
}

#[derive(Deserialize)]
struct Timings {
    predicted_n: Option<u64>,
    _predicted_ms: Option<f64>,
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct HealthResponse {
    status: Option<String>,
}

impl LlamaCppBackend {
    pub fn new(name: &str, model_id: &str, base_url: &str) -> Self {
        Self {
            name: name.to_string(),
            model_id: model_id.to_string(),
            base_url: base_url.trim_end_matches('/').to_string(),
            client: Client::builder()
                .timeout(Duration::from_secs(60))
                .build()
                .expect("valid reqwest client"),
        }
    }
}

#[async_trait]
impl ModelBackend for LlamaCppBackend {
    fn name(&self) -> &str {
        &self.name
    }

    fn model_id(&self) -> &str {
        &self.model_id
    }

    #[instrument(skip(self, prompt))]
    async fn generate(
        &self,
        prompt: &str,
        max_tokens: Option<i32>,
        temperature: Option<f64>,
    ) -> Result<GenerationResult, ModelError> {
        let start = Instant::now();

        let request = CompletionRequest {
            prompt: prompt.to_string(),
            n_predict: max_tokens,
            temperature,
            stream: false,
        };

        let resp = self
            .client
            .post(format!("{}/completion", self.base_url))
            .json(&request)
            .send()
            .await?;

        if !resp.status().is_success() {
            return Err(ModelError::BackendUnavailable(format!(
                "llama.cpp returned status {}",
                resp.status()
            )));
        }

        let body: CompletionResponse = resp.json().await?;
        let duration = start.elapsed().as_millis() as u64;

        Ok(GenerationResult {
            text: body.content,
            model_name: self.model_id.clone(),
            tokens_in: prompt.len() as u64 / 4,
            tokens_out: body.timings.as_ref().and_then(|t| t.predicted_n).unwrap_or(0),
            duration_ms: duration.max(1),
            error: None,
        })
    }

    async fn generate_stream(
        &self,
        prompt: &str,
        max_tokens: Option<i32>,
        temperature: Option<f64>,
    ) -> Result<Vec<String>, ModelError> {
        let request = CompletionRequest {
            prompt: prompt.to_string(),
            n_predict: max_tokens,
            temperature,
            stream: true,
        };

        let resp = self
            .client
            .post(format!("{}/completion", self.base_url))
            .json(&request)
            .send()
            .await?;

        if !resp.status().is_success() {
            return Err(ModelError::BackendUnavailable(format!(
                "llama.cpp returned status {}",
                resp.status()
            )));
        }

        let mut chunks = Vec::new();
        use futures::StreamExt;
        let mut stream = resp.bytes_stream();
        let mut buf = String::new();

        while let Some(chunk) = stream.next().await {
            let chunk = chunk?;
            buf.push_str(&String::from_utf8_lossy(&chunk));
            while let Some(line_end) = buf.find('\n') {
                let line: String = buf.drain(..=line_end).collect();
                let line = line.trim();
                if line.is_empty() {
                    continue;
                }
                if let Ok(partial) = serde_json::from_str::<serde_json::Value>(line) {
                    if let Some(content) = partial["content"].as_str() {
                        if !content.is_empty() {
                            chunks.push(content.to_string());
                        }
                    }
                    if partial["stop"].as_bool().unwrap_or(false) {
                        break;
                    }
                }
            }
        }

        Ok(chunks)
    }

    async fn health_check(&self) -> Result<bool, ModelError> {
        let resp = self
            .client
            .get(format!("{}/health", self.base_url))
            .send()
            .await?;

        Ok(resp.status().is_success())
    }
}
